be/src/connector/hive_connector.cpp

```diff
@@ -579,6 +579,95 @@ HdfsScanner* HiveDataSource::_create_paimon_jni_scanner(const FSOptions& options
+    auto* partition_desc = hdfs_table->get_partition(scan_range.partition_id);
+    std::string partition_full_path = partition_desc->location();
+
+    std::string required_fields;
+    for (auto const& slot : _materialize_slots) {
+        required_fields.append(slot->col_name());
+        required_fields.append(",");
+    }
+    required_fields = required_fields.substr(0, required_fields.size() - 1);
+
+    std::string nested_fields;
+    for (auto slot : _materialize_slots) {
+        const TypeDescriptor& type = slot->type();
+        if (type.is_complex_type()) {
+            build_nested_fields(type, slot->col_name(), &nested_fields);
+        }
+    }
+    if (!nested_fields.empty()) {
+        nested_fields = nested_fields.substr(0, nested_fields.size() - 1);
+    }
+
+    std::string data_file_path;
+    if (scan_range.relative_path.empty()) {
+        data_file_path = "";
+    } else {
+        data_file_path = fmt::format("{}/{}", partition_full_path, scan_range.relative_path);
+    }
+
+    std::map<std::string, std::string> jni_scanner_params;
+    jni_scanner_params["hive_column_names"] = hdfs_table->get_hive_column_names();
+    jni_scanner_params["hive_column_types"] = hdfs_table->get_hive_column_types();
+    jni_scanner_params["required_fields"] = required_fields;
+    jni_scanner_params["nested_fields"] = nested_fields;
+    jni_scanner_params["data_file_path"] = data_file_path;
+    jni_scanner_params["block_offset"] = std::to_string(scan_range.offset);
+    jni_scanner_params["block_length"] = std::to_string(scan_range.length);
+    jni_scanner_params["serde"] = hdfs_table->get_serde_lib();
+    jni_scanner_params["input_format"] = hdfs_table->get_input_format();
+    jni_scanner_params["fs_options_props"] = build_fs_options_properties(options);
+
+    std::string scanner_factory_class = "com/starrocks/hive/reader/HiveScannerFactory";
+
+    HdfsScanner* scanner = _pool.add(new JniScanner(scanner_factory_class, jni_scanner_params));
+    return scanner;
+}
+
+HdfsScanner* HiveDataSource::_create_file_table_jni_scanner(const FSOptions& options) {
+    const auto& scan_range = _scan_range;
+    const auto* file_table = dynamic_cast<const FileTableDescriptor*>(_hive_table);
+
+    std::string required_fields;
+    for (auto const& slot : _materialize_slots) {
+        required_fields.append(slot->col_name());
+        required_fields.append(",");
+    }
+    required_fields = required_fields.substr(0, required_fields.size() - 1);
+
+    std::string nested_fields;
+    for (auto slot : _materialize_slots) {
+        const TypeDescriptor& type = slot->type();
+        if (type.is_complex_type()) {
+            build_nested_fields(type, slot->col_name(), &nested_fields);
+        }
+    }
+    if (!nested_fields.empty()) {
+        nested_fields = nested_fields.substr(0, nested_fields.size() - 1);
+    }
+
+    std::map<std::string, std::string> jni_scanner_params;
```

Comment:
what's difference to hive table? I see a lot of duplicated code, can they be merged into one?

Reply:
done